storybookjs/storybook
# components: format Source code with Prettier

Issue: Relevant to https://github.com/storybookjs/storybook/issues/8078 (doesn't close), https://github.com/storybookjs/storybook/issues/8078#issuecomment-605357603, https://github.com/storybookjs/storybook/issues/8078#issuecomment-605430645

## What I did

- [x] Augmented formatter with `prettier/standalone`.
  - Parsers are imported with `import()` function, so they should be loaded at runtime only when needed.
- [x] Allowed passing SourceProps through `withSource` prop of Preview to opt-out of formatting.
  - Example:
    ```jsx
    <Preview
        withSource={{
            language: 'jsx',
            format: false,
            state: SourceState.OPEN
        }}
    >
      <Story name="unformatted JSX in MDX">{/* ... */}</Story>
    </Preview>
    ```


### Notes

- The tests are in: lib/components/src/syntaxhighlighter/formatter.test.js
- It's used only in: lib/components/src/syntaxhighlighter/syntaxhighlighter.tsx
- Prettier usually infers parser from filename.
In the browser, we have to pass it explicite.
Supported parsers include: `angular`, `vue`,  `babel`, `mdx`, `css`, `html`.
    - https://prettier.io/docs/en/options.html#parser
    - https://prettier.io/docs/en/browser.html#es-modules
    - Parsers are published as separate entry points in prettier package.
![](https://firebasestorage.googleapis.com/v0/b/firescript-577a2.appspot.com/o/imgs%2Fapp%2F2ndbrain%2F_6ETWNKDRH?alt=media&token=52739c65-77cc-42c4-8f25-cfdda44fe5c6)
    - I plan to add support for `babel`, `css` and `html` in this pull request.
I believe following up on this won't be hard.
- It's worth to note the significant bundle size of prettier/standalone (~**1mB unpacked**) and parsers (prettier-plugin-svelte is **3.4kB gzipped**, prettier-plugin-reasonml is **13kB gzipped**).
    - It might be beneficial to import `prettier/standalone` only when the format prop is set to `true`.
    - **Question:** Should storybook packages for particular framework (e.g. @storybook/vue) add parsers explicitly? Or should they be dynamically imported depending on the `language` prop passed to `Source`.

## How to test

- Is this testable with Jest or Chromatic screenshots?
  - I've tested my changes in `formatter.test.js` and added relevant stories to 
    - `syntaxhighlighter.stories.tsx`
    - `Source.stories.tsx`
   - `formatted-preview.stories.mdx`
     - These stories use the topmost API, so this might be a good place for Chromatic screenshots. I'm not familiar with Chromatic, so I'm not sure how to proceed further.
- Does this need a new example in the kitchen sink apps?
  - I'm not sure.
- Does this need an update to the documentation?
  - [x] A pararaph in [addons/docs/docs/mdx.md](https://github.com/storybookjs/storybook/blob/54757192c5bd8235c6343b4e0efff120fe61b5d4/addons/docs/docs/mdx.md) would be useful.

If your answer is yes to any of these, please make sure to include it in your PR.
